packages/push/src/genome/plushy.rs

```diff
@@ -1,68 +1,134 @@
 use crate::instruction::PushInstruction;
+use easy_cast::ConvApprox;
 use ec_core::{
     generator::{collection::CollectionGenerator, Generator},
     genome::Genome,
 };
 use ec_linear::genome::Linear;
-use rand::rngs::ThreadRng;
+use rand::{rngs::ThreadRng, Rng};
+
+#[derive(Clone, Eq, PartialEq)]
+pub enum PushGene {
+    Close,
+    Instruction(PushInstruction),
+}
+
+impl<T> From<T> for PushGene
+where
+    T: Into<PushInstruction>,
+{
+    fn from(instruction: T) -> Self {
+        Self::Instruction(instruction.into())
+    }
+}
+
+impl std::fmt::Debug for PushGene {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Close => write!(f, "Close"),
+            Self::Instruction(instruction) => instruction.fmt(f),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GeneGenerator {
+    close_probability: f32,
+    instructions: Vec<PushInstruction>,
+}
+
+impl GeneGenerator {
+    #[must_use]
+    pub fn new(close_probability: f32, instructions: Vec<PushInstruction>) -> Self {
+        Self {
+            close_probability,
+            instructions,
+        }
+    }
+
+    #[must_use]
+    pub fn with_uniform_close_probability(instructions: Vec<PushInstruction>) -> Self {
+        Self {
+            close_probability: 1.0 / f32::conv_approx(instructions.len() + 1),
+            instructions,
+        }
+    }
+}
+
+impl Generator<PushGene> for GeneGenerator {
+    fn generate(&self, rng: &mut ThreadRng) -> anyhow::Result<PushGene> {
+        if rng.gen::<f32>() < self.close_probability {
+            Ok(PushGene::Close)
+        } else {
+            self.instructions.generate(rng).map(PushGene::Instruction)
+        }
+    }
+}
 
 #[derive(Debug, Clone, Eq, PartialEq)]
 pub struct Plushy {
-    instructions: Vec<PushInstruction>,
+    genes: Vec<PushGene>,
 }
 
 // TODO: We might want to implement some sort of `Into`
 //   trait instead of just having a getter. Having something
 //   like `to_instructions()` since we're cloning?
 impl Plushy {
+    pub fn new<T: IntoIterator<Item = PushGene>>(iterable: T) -> Self {
```

Comment:
```suggestion
    pub fn new(iterable: impl IntoIterator<Item = PushGene>) -> Self {
```
We might as well use  a `impl T` here since Iterator types are hard to specify in general most of the time and should be inferred, as such a generic doesn't really have any advantage